share/source.F90

```diff
@@ -12,36 +12,98 @@
+!                particle possiblely located.
+!     cell_list_size: integer. Size of the cell_list.
+!
+! There are 3 types of not found icl_tmp (icl_tmp == -1):
+! - Type 1: Sorce particle is located in a cell that exist in neutron transport
+!             but removed in photon transport. Therefore, the cell number does
+!             not exist in the ncl list.
+!           This is not an error, happens with samll frequency.
+!           Skip it and resample next particle without warning message.
+! - Type 2: Source particle is not located in the given cell_list for HEX
+!             mesh cases (both voxel and sub-voxel). Because cell_list from
+!             discretize_geom() missed some cells with very small volume
+!             fractions.
+!           This is caused by random error, happens with samll frequency.
+!           Skip it andd resample next particle with warning message.
+! - Type 3: Source partilce with a specific coordinates can't be find in any
+!             cell from ncl(1) to ncl(mxa).
+!           This is an error. When this error happens, it means that there is
+!             something wrong in this file (source.F90) or DAGMC geometry
+!             representing/checking. It happens with low frenquency for some
+!             complex geometry. The results is suspicious under this condition.
+!           Skip it and resample next particle with error message.
+
 
     use mcnp_global
     use mcnp_debug
     ! xxx,yyy,zzz are global variables
     ! mxa is global
     integer :: i ! iterator variable
-    integer :: j ! tempory cell test
+    integer :: j ! temporary cell test
     integer :: icl_tmp ! temporary cell variable
+    integer, intent(in) :: cell_list_size
+    integer, dimension(cell_list_size), intent(in) :: cell_list
+    integer :: cid ! cell index
+
     icl_tmp = -1
+    ! If the cell_list is given (for HEX mesh),
+    ! use it to find cell first
+    if (cell_list_size > 0) then
+        ! HEX mesh. VOXEL/SUBVOXEL
+        do i = 1, cell_list_size
+           if (cell_list(i) .le. 0) then
+               ! not a valid cell number (-1)
+               exit
+           endif
+           cid = namchg(1, cell_list(i))
+           if (cid .eq. 0) then
+               ! Type 1: cell index not found, skip and resampling
+               exit
+           endif
+           call chkcel(cid, 0, j)
+           if (j .eq. 0) then
+              ! valid cell found
+              icl_tmp = cid
+              exit
+           endif
+        enddo
+    endif
 
-    do i = 1, mxa
-      call chkcel(i, 0, j)
-      if (j .eq. 0) then
-         ! valid cel set
-         icl_tmp = i
-         ! icl now set to be valid cell
-         exit
+    ! If the icl_tmp is not find yet (for HEX mesh), type 2 or type 3 happends,
+    ! or the cell_list is not given (for TET mesh),
+    ! find it in the entire list of cells
+    if (icl_tmp == -1) .or. (cell_list_size .eq. 0) then
```

Comment:
This line gives me compiler errors. Adding parentheses around the entire if statement; i.e.
```
if ((icl_tmp .eq. -1) .or. (cell_list_size .eq. 0)) then
```
fixes the errors.

Reply:
My bad, I should have tested the compilation before commit it.